### ISSUE TITLE
build issue

### ISSUE THREAD

**Issue by puffanddmx** [paritytech/ink#491]
for the flipper ink basic example test, if i do cargo +nightly contract build, then it will be failed in centos 7.

i add this to the #![feature(const_fn)] in lib.rs -> also same

what is the problem?

Compiling ink_lang v2.1.0 (https://github.com/paritytech/ink?tag=latest-v2#cca31543)
error[E0723]: mutable references in const fn are unstable
   --> /root/.cargo/git/checkouts/ink-1add513eda8f5a89/cca3154/lang/src/dispatcher.rs:186:30
    |
186 |               pub const fn new(dispatchable: $dispatchable_fn<Msg, S>) -> Self {
    |                                ^^^^^^^^^^^^
...
242 | / impl_dispatcher_for! {
243 | |     /// Dispatcher for potentially storage mutating messages and constructors.
244 | |     struct DispatcherMut(mut DispatchableFnMut);
245 | | }
    | |_- in this macro invocation
    |
    = note: see issue #57563 <https://github.com/rust-lang/rust/issues/57563> for more information
    = help: add `#![feature(const_fn)]` to the crate attributes to enable
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0723`.
error: could not compile `ink_lang`

To learn more, run the command again with --verbose.

**Comment by Robbepop** [comment]
Could you please paste your contract source code?

**Comment by puffanddmx** [comment]
https://substrate.dev/substrate-contracts-workshop/#/0/creating-an-ink-project
-> $ cargo contract new flipper
-> $ cargo +nightly contract build
-> it is not new one, i just test the ink default flipper above url on centos 7

**Comment by nmrshll** [comment]
Same error here, both on macOS and in Docker with an ubuntu image:

```shell
Compiling ink_lang v2.1.0 (https://github.com/paritytech/ink?tag=latest-v2#cca31543)
error[E0723]: mutable references in const fn are unstable
```

## Steps to reproduce

### on my machine (macOS mojave):
In a new folder, run the steps from [the tutorial](https://substrate.dev/substrate-contracts-workshop/#/0/setup):

```shell
curl https://getsubstrate.io -sSf | bash -s -- --fast
rustup target add wasm32-unknown-unknown --toolchain stable
rustup component add rust-src --toolchain nightly
rustup toolchain install nightly-2020-06-01
rustup target add wasm32-unknown-unknown --toolchain nightly-2020-06-01

cargo +nightly-2020-06-01 install node-cli --git https://github.com/paritytech/substrate.git --tag v2.0.0-rc4 --force --locked
cargo install cargo-contract --vers 0.6.2 --force

cargo contract new flipper
cd flipper/
cargo +nightly test
```

### OR: in docker:

```dockerfile
# ./dockerfile
FROM phusion/baseimage:0.11
RUN apt-get update && \
    apt-get dist-upgrade -y -o Dpkg::Options::="--force-confold" && \
    apt-get install -y pkg-config libssl-dev git
RUN curl https://sh.rustup.rs -sSf | sh -s -- -y
ENV PATH="/root/.cargo/bin:${PATH}"

RUN curl https://getsubstrate.io -sSf | bash -s -- --fast
RUN rustup target add wasm32-unknown-unknown --toolchain stable
RUN rustup component add rust-src --toolchain nightly
RUN rustup toolchain install nightly-2020-06-01
RUN rustup target add wasm32-unknown-unknown --toolchain nightly-2020-06-01

RUN cargo +nightly-2020-06-01 install node-cli --git https://github.com/paritytech/substrate.git --tag v2.0.0-rc4 --force --locked
RUN cargo install cargo-contract --vers 0.6.2 --force

RUN cargo contract new flipper
WORKDIR flipper/
RUN cargo +nightly test
```
Build the dockerfile above with `docker build .` (docker daemon must be running)


## What I tried
- running the steps above (on my machine)
- deleting then reinstalling rust and all toolchains with:
 ```
rustup self uninstall
curl https://sh.rustup.rs -sSf | sh -s -- -y
```
and then running the same steps above
- in docker, on an ubuntu-based image (to try a completely clean install with a different system)

**Comment by Kleak** [comment]
got same error on linux today

**Comment by Robbepop** [comment]
The line of code in question is this:
https://github.com/paritytech/ink/blob/v2.1.0/lang/src/dispatcher.rs#L186

Since this line (or any directly related code) is no longer contained in ink! 3.0 this will automatically be fixed as soon as ink! 3.0 is released, I assume.
Unfortunately it has been a long time since I was working with this old code but it might be that more recent nightly Rust versions implemented some new checks that might not be compatible with it which is why you see the errors recently.
A fix could be to try using an older nightly compiler from well before the incident and see if that fixes it and use that one until we finally release ink! 3.0 which is happening very soon.

**Comment by puffanddmx** [comment]
Thank you for answer.
Then, if you temporarily delete the function that has an error when building the cargo contract and build it, the error will not occur right ?

**Comment by Robbepop** [comment]
Since ink! codegen depends on this function simply removing it will probably yield other errors then. So would be a proper fix. The proper fix is the upcoming new codegen of ink! 3.0.

**Comment by puffanddmx** [comment]
yep right!
i test it , it depends of a lot of contstructor!!

when is the day when we can use new codegen of ink! 3.0 ?
are there any way that we can solve this before new codegen of ink! 3.0 ?

**Comment by nmrshll** [comment]
might be late to the party but here's a potential solution:

## TL;DR

I've put together a working flipper contract (with recent versions of rust and ink) at https://github.com/nmrshll-issues/paritytech-ink-491-build-issue

Tests, building the Wasm file and the ABI all work, so you can continue the tutorial from there.
There's a makefile to make sure you get the exact same installation process and versions as me.



## How to reproduce
(e.g. with a different example contract)

I went through all the [git revisions of `paritytech/ink`](https://github.com/paritytech/ink/compare/v2.1.0...master), found the revision `1e312d1` where examples build and CI checks pass.

From there, I copied the code of `examples/flipper/`.

And then just adapted the dependencies in Cargo.toml to point to the revision `1e312d1`. E.g. :
```toml
ink_primitives = { version = "2.1.0", path = "../../primitives", default-features = false }
ink_abi = { version = "2.1.0", path = "../../abi", default-features = false, features = ["derive"], optional = true }
ink_core = { version = "2.1.0", path = "../../core", default-features = false }
ink_lang = { version = "2.1.0", path = "../../lang", default-features = false }
```
becomes:
```toml
ink_primitives = {git="https://github.com/paritytech/ink",rev="1e312d1",default-features = false }
ink_abi = {git="https://github.com/paritytech/ink",rev="1e312d1", default-features = false, features = ["derive"], optional = true }
ink_core = {git="https://github.com/paritytech/ink",rev="1e312d1",default-features = false }
ink_lang = {git="https://github.com/paritytech/ink",rev="1e312d1",default-features = false }
```


## Caveats
- the syntax for contracts has changed a bit with recent versions of ink (compared to the tutorial)
- the ABI builds but doesn't seem to upload to the node (with the current version of [`polkadot-js/apps`](https://github.com/polkadot-js/apps))

**Comment by Robbepop** [comment]
I recommend you to wait for our ink! 3.0 release.
We also will provide a UI that works out-of-the-box with the new metadata format of ink!.
Also we are going to update our ink! workshop for the upcoming release.
Stay tuned. Won't take too long until then.

**Comment by nmrshll** [comment]
Confirmed, `cargo-contract 0.7` just works great out of the box with recent versions of rust nightly
Thanks for the demo of it at sub0 @Robbepop (didn't realize it was you until just now)
For everyone else here it is: https://www.crowdcast.io/e/axvfinsv/20
